lib/core/connection/connect.js

```diff
@@ -96,77 +74,115 @@ function performInitialHandshake(conn, options, _callback) {
+      // The handshake technically is a monitoring check, so its socket timeout should be connectTimeoutMS
+      handshakeOptions.socketTimeout = options.connectTimeoutMS || options.connectionTimeout;
     }
 
-    const supportedServerErr = checkSupportedServer(ismaster, options);
-    if (supportedServerErr) {
-      callback(supportedServerErr);
-      return;
-    }
+    const start = new Date().getTime();
+    conn.command('admin.$cmd', handshakeDoc, handshakeOptions, (err, result) => {
+      if (err) {
+        callback(err);
+        return;
+      }
 
-    if (!isModernConnectionType(conn)) {
-      // resolve compression
-      if (ismaster.compression) {
-        const agreedCompressors = compressors.filter(
-          compressor => ismaster.compression.indexOf(compressor) !== -1
-        );
+      const response = result.result;
+      if (response.ok === 0) {
+        callback(new MongoError(response));
+        return;
+      }
 
-        if (agreedCompressors.length) {
-          conn.agreedCompressor = agreedCompressors[0];
-        }
+      const supportedServerErr = checkSupportedServer(response, options);
+      if (supportedServerErr) {
+        callback(supportedServerErr);
+        return;
+      }
+
+      if (!isModernConnectionType(conn)) {
+        // resolve compression
+        if (response.compression) {
+          const agreedCompressors = handshakeDoc.compression.filter(
+            compressor => response.compression.indexOf(compressor) !== -1
+          );
 
-        if (options.compression && options.compression.zlibCompressionLevel) {
-          conn.zlibCompressionLevel = options.compression.zlibCompressionLevel;
+          if (agreedCompressors.length) {
+            conn.agreedCompressor = agreedCompressors[0];
+          }
+
+          if (options.compression && options.compression.zlibCompressionLevel) {
+            conn.zlibCompressionLevel = options.compression.zlibCompressionLevel;
+          }
         }
       }
-    }
 
-    // NOTE: This is metadata attached to the connection while porting away from
-    //       handshake being done in the `Server` class. Likely, it should be
-    //       relocated, or at very least restructured.
-    conn.ismaster = ismaster;
-    conn.lastIsMasterMS = new Date().getTime() - start;
+      // NOTE: This is metadata attached to the connection while porting away from
+      //       handshake being done in the `Server` class. Likely, it should be
+      //       relocated, or at very least restructured.
+      conn.ismaster = response;
+      conn.lastIsMasterMS = new Date().getTime() - start;
 
-    const credentials = options.credentials;
-    if (!ismaster.arbiterOnly && credentials) {
-      authenticate(conn, credentials.resolveAuthMechanism(ismaster), callback);
+      if (!response.arbiterOnly && credentials) {
+        // store the response on auth context
+        Object.assign(authContext, { response });
```

Comment:
This kind of side-effect / mutation of state scares me, perhaps it's an irrational fear. I'd prefer some kind of more stream-lined obvious way of doing this that the parent class itself knows about. eg a setter `authContext.response = response` or `authContext.addResponse(response)`. I'd love to having typing somehow on `authContext`  that says it has `{ response?: any }`.